design/open/hard-delete.md

```diff
@@ -0,0 +1,400 @@
+- The S3 gateway `put` operation with the copy support will use the underlying adapter to copy the data. Will require from each adapter to support/emulate a 'copy' operation.
+* (optional) Enable 'move' API as alternative to 'copy' we have seen that moving objects from one location to another by copy+delete of metadata will enable easy support for lakeFSFS move. This can be implemented by marking the entry on staging as 'locked'. During the move operation - lock+copy+delete+unlock operation. In case of any failure, as we don't have transactions, we may keep two entries, or keep locked object without ever delete its physical address.
+
+TODO(Barak): list more cases and how it will be address in this solution.
+
+### Flows
+
+#### Move operation
+1. Get entry
+2. "Lock" entry (`SetIf`)
+3. Copy entry to new path on current staging token
+4. Delete old entry
+5. "Unlock" entry (new entry) 
+
+#### Upload Object
+
+1. Get entry
+2. "Lock" entry if exists (override scenario) (`SetIf`)
+3. Write blob
+4. Add staged entry (or update)
+5. Delete physical address if previously existed
+
+#### Delete Object
+1. Get entry
+2. "Lock" entry (`SetIf`) to protect from delete - move race
+3. Delete staging entry
+4. Hard delete object
+
+### Races and issues
+
+#### 1. Commit - Move Race
+1. Start Move:
+   1. Get entry
+   2. "Lock" entry
+2. Start Commit:
+   1. Change staging token
+   2. write commit data (old entry path was written)
+3. Move:
+   1. Create new entry on new staging token
+   2. Delete old entry
+   3. "Unlock" physical address
+
+Now we have an uncommitted entry pointing to a committed physical address. If we reset the branch we will delete
+committed data!
+
+#### 2. Delete - Move Race
+1. Start Delete:
+   1. Get entry
+   2. Check if not "locked"
+2. Start Move:
+   1. Get entry
+   2. "Lock" entry
+   3. Copy entry to new path on current staging token
+3. Delete:
+   1. Hard Delete object from store
+
+In this situation we have an uncommitted entry which points to an invalid physical address.
+**Resolved by locking the entry on delete flow as well**
+
+#### 3. Concurrent Move
+Concurrent move should be blocked as it may cause multiple pointers for the same physical address.
+The most straightforward way to do so is to rely on the absence of the lock.
+The problem with this approach is that permanently locked addresses (which is a protection mechanism in case of catastrophic failures)
+can never be moved -which leads us to the stale lock problem.
+
+#### 4. Stale lock problem
+We use the lock mechanism to protect from race scenarios on Upload (override), Move and Delete. As such, an entry with a stale
+lock will prevent us from performing any of these operations.
+We can overcome it with a time based lock - but this might present additional challenges to the proposed solution.
```

Comment:
How can we not have it? Can we live with a system that some entries are locked for good?

Reply:
IMHO this is not something that we can live with. But - we need to take into account that using a time based lock might introduce new problems

Reply:
So let's break it down then

Reply:
As I see it - we should first try to solve issue #1 (Commit - Move Race), before taking the time to solve this one.
Seems like the commit move race is our blocker.

Reply:
Time based locks are inherently unsafe unless your infrastructure has fairly strict guarantees on clocks.  We need to ensure that at least AWS + EKS provide such a guarantee.